[20044] Upgrade to Rails 4.2.x

myabc opened the pull request:
:warning: **Please review, merge opf/openproject-translations#29 first**, adjusting branches in `Gemfile` as appropriate.
:warning: ~~**Includes commits from #3382. Please review, merge first.**~~

https://community.openproject.org/work_packages/20044

basiszwo committed:
Use factory girl instead of object daddy

basiszwo committed:
Remove object daddy and its exemplars

basiszwo committed:
Remove object daddy requirement for cukes

basiszwo committed:
Decouple legacy specs from ObjectDaddy

myabc committed:
Bump Rails to 4.2.4

Signed-off-by: Alex Coles <alex@alexbcoles.com>

myabc committed:
Switch to 4.2 branch of prototype-rails

prototype-rails is no longer officially supported for Rails 4.1+

Signed-off-by: Alex Coles <alex@alexbcoles.com>

basiszwo committed:
Use activejob deliver_now method (removes deprecation warning)

myabc committed:
Rename Rails Dom Testing constant

Signed-off-by: Alex Coles <alex@alexbcoles.com>

TeatroIO commented:
I've prepared a stage to preview changes. <a href='http://myabc-feature-rails4-2.opf-openproject-2af22bf12c7dfdad4ddf.ttrcloud.com' target='_blank'>Open stage</a> or <a href='http://myabc-feature-rails4-2.opf-openproject-2af22bf12c7dfdad4ddf.ttrcloud.com/teatro.log' target='_blank'>view logs</a>.

basiszwo commented:
- [ ] remove deprecation warning about `sanitize_sql_hash_for_conditions` by upgrading act_as_list. see https://github.com/swanandp/acts_as_list/pull/140

myabc committed:
Bump Globalize to 5.0.x (AR 4.2-compatible)

Signed-off-by: Alex Coles <alex@alexbcoles.com>

myabc committed:
Update MenuHelper legacy spec for Nokogiri switch

Rails Dom Testing now uses Nokogiri.

Signed-off-by: Alex Coles <alex@alexbcoles.com>

myabc committed:
Add Responders dependency

Bring back `respond_with` support, removed in Rails 4.2.

Signed-off-by: Alex Coles <alex@alexbcoles.com>

myabc committed:
Fix DOM testing in legacy layout spec

Signed-off-by: Alex Coles <alex@alexbcoles.com>

myabc committed:
Switch to new AR API for boolean typecasting

Signed-off-by: Alex Coles <alex@alexbcoles.com>

ulferts commented on app/controllers/account_controller.rb:
```diff
@@ -100,7 +100,7 @@ def lost_password
         token = Token.new(user: user, action: 'recovery')
         if token.save
-          UserMailer.password_lost(token).deliver
+          UserMailer.password_lost(token).deliver_now
```
I realize, this is the 1:1 translation but can't we use `deliver_later` instead @basiszwo, @myabc? Would this use our readily available delayed_job then?

myabc replied:
@ulferts sounds like a good idea, but perhaps better as a separate PR?

ulferts commented:
@myabc would you be so kind to rebase this on the current dev now that #3382 is merged?

myabc committed:
Merge branch 'dev' into feature/rails4.2

Signed-off-by: Alex Coles <alex@alexbcoles.com>

myabc committed:
Update Job spec for call to #deliver_now

See 868d764a.

Signed-off-by: Alex Coles <alex@alexbcoles.com>

myabc committed:
Replace further deprecated calls to #deliver

See 868d764a.

Signed-off-by: Alex Coles <alex@alexbcoles.com>

myabc committed:
Update expected HTML in FormTagHelper spec

Signed-off-by: Alex Coles <alex@alexbcoles.com>

myabc committed:
"De-legacy-fy" CustomFieldsHelper spec

* Improve HTML matching, to ignore order of HTML attributes.

Signed-off-by: Alex Coles <alex@alexbcoles.com>

myabc committed:
Fix ApplicationHelper legacy spec HTML matching

Ignore order of HTML attributes.

Signed-off-by: Alex Coles <alex@alexbcoles.com>

myabc committed:
Fix Textile::Formatter legacy spec HTML matching

Ignore order of HTML attributes.

Signed-off-by: Alex Coles <alex@alexbcoles.com>

myabc committed:
Fix deprecated Model.exists? arg in UpdateContract

Signed-off-by: Alex Coles <alex@alexbcoles.com>

myabc committed:
Update openproject-translations

Signed-off-by: Alex Coles <alex@alexbcoles.com>

myabc committed:
Merge branch 'dev' into feature/rails4.2

Signed-off-by: Alex Coles <alex@alexbcoles.com>

myabc committed:
Override Tableless.column to support AR 4.2

Pending compatible activerecord-tableless release.
See: https://github.com/softace/activerecord-tableless/issues/22

Signed-off-by: Alex Coles <alex@alexbcoles.com>

myabc committed:
Fix deprecated .find arg in Project

Signed-off-by: Alex Coles <alex@alexbcoles.com>

myabc commented on app/models/project.rb:
```diff
@@ -860,13 +860,18 @@ def add_work_package(attributes = {})
-      i.type = if type_attribute
-                 project.types.find(type_attribute)
+      type    = attributes.delete(:type)
+      type_id = if type && type.respond_to?(:id)
```
@ulferts am I being overly defensive here?

basiszwo replied:
In which case is it possible that type does not respond to `:id` ?

myabc replied:
@basiszwo I'm not that familiar with the `WorkPackage#add_work_package` API - I think this is a Redmine remnant. It's conceivable though that a user passes in an attributes `Hash` with a `type: nil`, e.g. `type: Type.find(by_name: 'unknown')`

myabc committed:
Bump acts_as_list to 0.7.x (AR 4.2-compatible)

Signed-off-by: Alex Coles <alex@alexbcoles.com>

myabc committed:
Fix broken #assert_select_rjs in Controller specs

Add workaround for nested assertions, broken since Rails Dom Testing
switch to Nokogiri.

Signed-off-by: Alex Coles <alex@alexbcoles.com>

myabc committed:
Fix invalid URLs in legacy integration specs

Signed-off-by: Alex Coles <alex@alexbcoles.com>

myabc committed:
Temp. mark failing Activity spec pending

Signed-off-by: Alex Coles <alex@alexbcoles.com>

myabc committed:
Update config for serve_static_files renaming

Signed-off-by: Alex Coles <alex@alexbcoles.com>

myabc committed:
Update binstubs

Signed-off-by: Alex Coles <alex@alexbcoles.com>

myabc committed:
Update config to raise in transactional callbacks

Signed-off-by: Alex Coles <alex@alexbcoles.com>

myabc committed:
Update Rails configs: boot.rb

Signed-off-by: Alex Coles <alex@alexbcoles.com>

myabc committed:
Fix for ActiveRecord::Base.reflections API change

`WorkPackage.reflections.keys` now returns Strings rather than
symbols.

This solution is not ideal: calling `to_sym` is often a code smell.
It might be acceptable to pass through Strings to `includes`, but
this warrants further discussion.

Signed-off-by: Alex Coles <alex@alexbcoles.com>

myabc committed:
Replace anonymous class in ActsAsOpEngine spec

Rails 4.2 is not able generate an appropriate `railtie_name` (aliased
to `engine_name`) for anonymous classes - this, in turn, breaks
initializer definitions.

Signed-off-by: Alex Coles <alex@alexbcoles.com>

myabc committed:
Fix TabularFormBuilder HTML matching

Use #be_html_eql matcher in TabularFormBuilder spec (see 1f42f127)
to avoid relying on specific order of attributes.

Signed-off-by: Alex Coles <alex@alexbcoles.com>

myabc committed:
Temp. switch to rspec-activemodel-mocks from git

Pending release of 1.0.2, including the following fixes:
https://github.com/rspec/rspec-activemodel-mocks/pull/19

Signed-off-by: Alex Coles <alex@alexbcoles.com>

myabc committed:
Remove ObjectDaddy remnants from cuke steps

See d2edc0d0.

Signed-off-by: Alex Coles <alex@alexbcoles.com>

myabc commented:
@ulferts @basiszwo @NobodysNightmare @oliverguenther we're getting closer to green CI. In order to keep things on track, I'd really appreciate some help/guidance in two areas:
- [x] ~~`Setting` callbacks no longer seem to work (at least in specs). This also causes `AvatarHelper` specs to fail.~~ https://travis-ci.org/opf/openproject/jobs/80118468#L2578
- [x] `Activity`: https://travis-ci.org/opf/openproject/jobs/80118468#L2653 I also marked a legacy spec as pending. 97b24b94533ec9577867590c6c2fd251476f2501 - @ulferts does this look like a permissions issue?

myabc committed:
Fix Settings callbacks for ActiveSupport API change

Identify callbacks with String not Symbol.

Signed-off-by: Alex Coles <alex@alexbcoles.com>

myabc committed:
Merge branch 'dev' into feature/rails4.2

Signed-off-by: Alex Coles <alex@alexbcoles.com>

oliverguenther commented:
> Setting callbacks no longer seem to work (at least in specs). 

This appears to be due to the symbol that is passed to `instrument`, which appears to now be used directly rather than stringified in the AS instrumenter. Changing the call to use a string at all times seems to fix it.

/Edit: Ah. I see you just addressed this yourself. Never mind then.

I've opened https://github.com/myabc/openproject/pull/8 for it, going to push more when I find it.

myabc commented:
> /Edit: Ah. I see you just addressed this yourself. Never mind then.

@oliverguenther sorry for the duplicated effort (you're too fast, or I'm too slow :wink:).

myabc committed:
Use #be_html_eql matcher in TextFormatting spec

Signed-off-by: Alex Coles <alex@alexbcoles.com>

myabc committed:
Use #be_html_eql matcher in APIv3 Render spec

Signed-off-by: Alex Coles <alex@alexbcoles.com>

myabc committed:
Use #be_html_eql matcher in ApplicationHelper spec

Signed-off-by: Alex Coles <alex@alexbcoles.com>

myabc committed:
Use #be_html_eql matcher in AvatarHelper spec

Signed-off-by: Alex Coles <alex@alexbcoles.com>

myabc committed:
Merge branch 'dev' into feature/rails4.2

Signed-off-by: Alex Coles <alex@alexbcoles.com>

ulferts commented:
> @ulferts does this look like a permissions issue?

@myabc I'll take a look at it once I am done with my current review.

myabc committed:
Re-remove deprecated call to #deliver

Signed-off-by: Alex Coles <alex@alexbcoles.com>

myabc committed:
Update expected HTML in RevisionRepresenter spec

Signed-off-by: Alex Coles <alex@alexbcoles.com>

ulferts committed:
fix project permission condition in activity

Without the #flatten, the arel visitor casts the two arrays to NULL.

ulferts commented:
@myabc I assume to have fixed the [failing activity spec](https://github.com/opf/openproject/pull/3421#issuecomment-139910923)  in myabc#9. There seems to be a change in the arel behaviour when you pass nested arrays to `in`. It apparently used to be flattened but now the arel visitor will transform it each array to NULL

myabc committed:
Merge pull request #9 from ulferts/fix/legacy_activity_spec

fix project permission condition in activity

myabc committed:
Fix, improve robustness of Types admin cuke

Signed-off-by: Alex Coles <alex@alexbcoles.com>

myabc committed:
Fix deprecated call to #deliver in UserMailer spec

Signed-off-by: Alex Coles <alex@alexbcoles.com>

myabc committed:
Merge branch 'dev' into feature/rails4.2

Signed-off-by: Alex Coles <alex@alexbcoles.com>

myabc committed:
Remove reundant include in MembersController spec

Signed-off-by: Alex Coles <alex@alexbcoles.com>

myabc commented on features/types/types_adminstration.feature:
```diff
@@ -90,4 +90,4 @@ Feature: Type Administration
   Scenario: A standard type cannot be deleted
      When I go to the global index page of types
      Then "Standard" should be the last element in the list
```
Note: `first element`/`last element` steps are actually broken (they don't check position in the list). Should be fixed here and in other places, but out-of-scope for this PR probably.

myabc commented:
@ulferts @basiszwo @NobodysNightmare @oliverguenther core is now ready for review. Note: I'm seeing some flickering with `user_mailer_spec`. Not sure if this also affects Rails 4.1/`dev`.

myabc committed:
Merge branch 'dev' into feature/rails4.2

Signed-off-by: Alex Coles <alex@alexbcoles.com>

myabc committed:
Update openproject-translations

Signed-off-by: Alex Coles <alex@alexbcoles.com>

myabc committed:
Add path helper shim to support :only_path option

`:only_path` is deprecated.

Signed-off-by: Alex Coles <alex@alexbcoles.com>

myabc committed:
Merge branch 'dev' into feature/rails4.2

Signed-off-by: Alex Coles <alex@alexbcoles.com>

myabc commented:
@oliverguenther any idea why the Repository features, in particular, might be flickering? https://travis-ci.org/opf/openproject/jobs/80581422

oliverguenther commented:
@myabc We are seeing this independent of the rails. 
I've tried to fix with this one on the CI several times, and I remember @ulferts having a go at it, too. I'll try to refactor the functionality, as it includes a full JS redirect and there seems to be some bug finding elements afterwards.

myabc commented:
@oliverguenther ok, thanks for clarification!

myabc commented:
@ulferts @basiszwo @NobodysNightmare @oliverguenther core + plugins are now ready for review (unfortunately I wasn't able to get a completely green build on Travis)

oliverguenther commented:
Changes in this PR look good to me, going to let the various configs run through.

ulferts commented:
Removing the WIP label, then. I hope to get a look at it today @myabc.

oliverguenther commented:
/Edit: Hmm,.. is working now. I blame `db:test:prepare`, although I ran this on a clean database...

myabc commented:
@oliverguenther what was the issue?

oliverguenther commented:
@myabc I was seeing numerous primary key errors on the myproject config upon saving / updating work packages, both in specs and on the UI.

```
ActiveRecord::RecordNotUnique (PG::UniqueViolation: ERROR:  duplicate key value violates unique constraint "work_package_journals_pkey"
DETAIL:  Key (id)=(2) already exists.
: INSERT INTO "work_package_journals" ("id", "type_id", "project_id", "subject", "description", "status_id", "priority_id", "author_id", "journal_id") VALUES ($1, $2, $3, $4, $5, $6, $7, $8, $9) RETURNING "id"):
  app/models/journal.rb:130:in `save_data'
  lib/plugins/acts_as_journalized/lib/redmine/acts/journalized/save_hooks.rb:95:in `save_journal_with_retry'
  lib/plugins/acts_as_journalized/lib/redmine/acts/journalized/save_hooks.rb:73:in `block in save_journals'
  lib/plugins/acts_as_journalized/lib/redmine/acts/journalized/save_hooks.rb:72:in `each'
  lib/plugins/acts_as_journalized/lib/redmine/acts/journalized/save_hooks.rb:72:in `save_journals'
  app/services/update_work_package_service.rb:46:in `save'
  lib/api/v3/work_packages/work_packages_api.rb:80:in `block (3 levels) in <class:WorkPackagesAPI>'
  app/middleware/params_parser_with_exclusion.rb:38:in `call'
```

oliverguenther commented:
Now I'm seeing the two following expected errors:

~~`can't modify frozen ActiveSupport::HashWithIndifferentAccess`
`silent_list` uses `changed_attributes[key] = foo`, which appears to be invalid in Rails 4.2:~~

Darn, you already fixed this and I've not checked out the 4.2 branch on silent_list. Never mind!

```
     Failure/Error: let(:type) { FactoryGirl.create(:type_standard) }
     ActiveRecord::RecordInvalid:
       Validation failed: Name has already been taken.
```

ulferts commented:
This s looking good to merge. Thanks for the tremendous job @myabc 

Will merge once the configurations are healthy again. They got broken by some other merges and I want to separate potential test breakages.

myabc commented:
:horse:
